### ISSUE TITLE
Issue using blank file_type in :upload_file

### ISSUE THREAD

**Issue by baseless** [devlab-oy/sepa#59]
Hi!
It seems that sepafm do not accept empty file_types for the :upload_file operation.
Since DanskeBank utilizes blank file_type as a valid format we have a need to get this working.

Example of their formats below: 

```
UMTL Global Transfer (SWIFT MT101). – only used in Finland
Must be blank EDIFACT 96 A format: PAYMUL
Must be blank EDIFACT 96A format: DIRDEB
Must be blank Danske Bank commaseparated format: Payment files
CMBO,CMUO,CMGB,CMBODK,CMSI,CMNI,CMFI,CMDE,CMGBDC,CMII,
CMUOGB,CMIIDC,CMUOIE,CMUO10,CMUOSE,CMUOFI,CMUODE,CMUONO
Must be blank Danske Bank commaseparated format: Collection files
COLLECTION,COLDEBMOD,COLDELETE
```
The error i receive when setting `file_type="" ` is "File type needs to be present and under 35 characters"

I did a check into your source and confirmed that in your code the following check is done:

```
def check_file_type
      if file_type.present?
        valid = file_type.size < 35
      else
        return if bank == :op && %i(download_file download_file_list).include?(command)
```
Tried inserting both "" and " " as file_type but i guess that it requires some form of content as it is built now? Do you have a workaround (Like som special notation if empty type) to get this working in current version? Or is this an issue that needs to be fixed? Thanks a lot in advance for your assistance!

**Comment by baseless** [comment]
I tried overriding this check by always returning true and if i do that the request goes through but i receive schema validation failed in the response. My guess is that this is caused by the <file_type></file_type> is skipped if the parameter is empty, but i havent had the time to verify this yet. But i just wanted to lift that adjusting the check_file_type method is not enough in itself.

**Comment by jonikanerva** [comment]
Thanks for the report! 

Haven't yet tried this out myself, but I think Nokogiri saves empty tags by default as `<file_type/>` instead of `<file_type ></file_type>`. This *might* cause problems in the Danske schema validation.

This behaviour can be changed with the [SaveOptions](http://www.rubydoc.info/github/sparklemotion/nokogiri/Nokogiri/XML/Node/SaveOptions). Could you try saving the xml with `save_with: Nokogiri::XML::Node::SaveOptions::NO_EMPTY_TAGS`? Maybe that could help.

**Comment by baseless** [comment]
Sounds like that could very well be the issue. Will see if i can try this workaround during the day. Just have to locate where to add this option (i guess should be set somewhere in soap_builder.rb or soap_danske.rb?).

**Comment by jonikanerva** [comment]
You can pass the `save_with` -option to the `to_xml` -method as a parameter. So maybe changing [this](https://github.com/devlab-oy/sepa/blob/f9b137ddfb86ea319007ebd5062bba6b2c4639f7/lib/sepa/application_request.rb#L36) to

```ruby
@application_request.to_xml save_with: Nokogiri::XML::Node::SaveOptions::NO_EMPTY_TAGS
```

could work..

I'll probably have some time this evening to look into this as well. 

I think we could remove the requirement that `upload_file` needs to have the file type set. Then it would be up to the user to make sure they are creating requests that are accepted by the banks.

**Comment by baseless** [comment]
Just wanted to mention that i tried the above suggestion and it seems like i wasnt enough to get it running.  Still seems to get `<ResponseCode>12</ResponseCode><ResponseText>Schema validation error.</ResponseText>` (i think i got failed, not error, before the nokogiri adjustment).

To summarize, there still seems to be some issue with the request xml format.

**Comment by baseless** [comment]
@jonikanerva did you find the time to look at this?

**Comment by baseless** [comment]
I did some more debugging on this and it looks like that when applying NO_EMPTY_TAGS it still works as long as a file type is actually set. NO_EMPTY_TAGS does however make CanonicalizationMethod non self closing as well so it affects other parts as well. 
When running this with an empty filetype the xml looks ok, seen to that the filetype is no longer self closing. But still get XML errors. This makes me suspect the real villain in this issue is hiding elsewhere in the code? Since everything looks ok when debugging the to_xml method but it still refuses to work.

By the way, i also tried just doing a .sub("<FIleType/>", "<FileType></FileType>") as well after the to_xml but that didnt work either.

**Comment by jonikanerva** [comment]
Thank you for the debugging effort. 

This can be a hard problem to solve without the help of Danske Bank. Solving this by trial and error could take a long time, as the soap request must be identical to the version Danske creates on their end. The soap request is canonicalized, which should in theory create identical files.

Did Danske provide you with an example of a working application request for uploading this file type? Or are you able to manually generate a valid request file? This would make debugging easier.

**Comment by baseless** [comment]
No i have not gotten a raw xml sample request from them, but as long as we enter some form of filetype (no matter what it is), it works in the the sense as we get a 00. And according to a rep for Danske that i talked to, it should work fine as long as we blank the file type for the targeted files. Which is the state in which we now reside.

I could talk to Danske again and ask for this and if he can see in more detail what goes wrong when sending blank type (not sure he can see that).

But the reason i was thinking about if it could be an issue with the digest / signature was the following:

If i set the parameter `file_type: 'TEST'` for example it works. BUT if i set parameter file_type: 'TESTABC' and then i edit the to_xml method as below (the only thing i actually add is the .sub) it will give schema validation error:

```
def to_xml
      set_common_nodes
      set_nodes_contents
      process_signature
      xmlReq = @application_request.to_xml save_with: Nokogiri::XML::Node::SaveOptions::NO_EMPTY_TAGS
      xmlReq = xmlReq.sub('TESTABC', 'TEST')
      xmlReq
    end
```

Havent checked this much deeper so far but as seen in the method above, process_signature is run befor turning the application_request to an XML string. This method both generates a digest of the request (INCLUDING the filetype tag) and then creates signature based on this. So my thought was that since it seems to digest the whole tag, maybe this causes the issue for the change when using NO_EMPTY_TAG as well?
So that generating a non-self-closing file_type tag makes the signature / digest no longer match the payload -> causing a xml schema validation error? (this is just a wild guess based on what i see in the code, but i wanted to mention it).

**Comment by jonikanerva** [comment]
Yes, you are correct! That was not the correct place to change the XML options. It should be done in the initializer.

If you modify the request *after* you've calculated the signature (as you do above), you should get an error. That is what the signature is for. It guarantees that the file content is not modified after it's signed.

**Comment by baseless** [comment]
Great then i got it correctly :)

But that leads to the next question ;), if this change -> `@application_request.to_xml` to `@application_request.to_xml save_with: Nokogiri::XML::Node::SaveOptions::NO_EMPTY_TAGS` then hypothetically causes the digest not to match the actual output when entering a blank file_type, the question is how to adjust it to ensure that the digest match the outputted xml string content. 

As i said above, not sure about this, just speculating at this point (but it seems like the most likely cause for this) :).

**Comment by jonikanerva** [comment]
I just [wrote a small test](https://github.com/devlab-oy/sepa/commit/a6639a55ea9239d705aba03a9ce97b3a6238c204) to see how the canonilization affects the file. The canonilized version of the XML *does* include the empty tags. So the signature is calculated from a file that contains the closing tags. Ie:

```xml
<ApplicationRequest xmlns=\"http://bxd.fi/xmldata/\">
  <CustomerId>360817</CustomerId>
  <Command>UploadFile</Command>
  <Timestamp>2016-12-05T14:46:41Z</Timestamp>
  <Environment>TEST</Environment>
  <SoftwareId>Sepa Transfer Library version 1.1.8</SoftwareId>
  <FileType></FileType>
  <Content>YTJsemMyRT0K
</Content>
```

This should match what Danske is expecting in their end. So I don't think the problem is in the XML representation. It must be something else.

**Comment by baseless** [comment]
I noticed that to when i debugged more now. Seems like the canonilization use correct type tag no matter the to_xml setting. Im a bit green in regards to ruby so unsure atm how to best /easiest verify the signature and digest of the resulting request soap object. But was my thought in regards to really ensure that there isnt an issue here. 

Anyway i plan to call my conact at danske tomorrow and ask him if he can see anything in their logs that can give more details on the issue than just the response code.

**Comment by baseless** [comment]
Sorry for the delay, took a while to get ananswer to this, but i have some information now. The information available was limited but the error message i got was as follows: 
`Schema validation error..http://1............/: cvc-simple-type 1: element {http://bxd.fi/xmldata/}FileType with anonymous type may not be empty.. `

If Danske runs the file on their side it works fine with empty file type. This seems not like the cause of a bug but a xml format issue?

**Comment by baseless** [comment]
I think i perhaps found a solution to this (just maybe). After having failed to get it working following a lot of debugging i did a simple hack to remove the tag completely. And then i got an OK. So awaiting confirmation if this actually gave the correct type in the end. If so ill get back with what i did.
